In [26]:
import pandas as pd
pd.options.display.max_columns = 20
pd.options.display.max_rows = 2500

In [27]:
df = pd.read_csv('basic_players_info.csv') # Load our extracted info
df2 = pd.read_csv('player_individual_stats.csv') # Load our extracted info
df.head()

name nationality         team  maps  rounds  KD-diff    KD  rating  \
0     ZywOo      France     Vitality    10     277       92  1.60    1.39   
1        iM     Romania  GamerLegion    12     306       68  1.34    1.35   
2     Spinx      Israel     Vitality    10     277       50  1.30    1.24   
3       NAF      Canada       Liquid     9     245       32  1.22    1.24   
4  YEKINDAR      Latvia       Liquid     9     245       30  1.19    1.20   

                       event  event_ID                     Full name  \
0  BLAST.tv Paris Major 2023      6973       Mathieu 'ZywOo' Herbaut   
1  BLAST.tv Paris Major 2023      6973               Mihai 'iM' Ivan   
2  BLAST.tv Paris Major 2023      6973          Lotan 'Spinx' Giladi   
3  BLAST.tv Paris Major 2023      6973          Keith 'NAF' Markovic   
4  BLAST.tv Paris Major 2023      6973  Mareks 'YEKINDAR' Gaļinskis   

   Player ID                                   Full link  
0      11893     https://www.hltv.org/player/11893/zywoo  
1      14759        https://www.hltv.org/player/14759/im  
2      18221     https://www.hltv.org/player/18221/spinx  
3       8520        https://www.hltv.org/player/8520/naf  
4      13915  https://www.hltv.org/player/13915/yekindar

In [28]:
df2.head()

name  Player ID  event_ID  Kills  Deaths  Kill / Death  Kill / Round  \
0     ZywOo      11893      6973    160      95          1.68          0.84   
1        iM      14759      6973    209     239          0.87          0.64   
2     Spinx      18221      6973    135     107          1.26          0.71   
3       NAF       8520      6973    181     174          1.04          0.66   
4  YEKINDAR      13915      6973    152     184          0.83          0.55   

   Rounds with kills  Kill - Death difference  Total opening kills  ...  \
0                 99                       65                   25  ...   
1                142                      -30                   44  ...   
2                 98                       28                   23  ...   
3                128                        7                   20  ...   
4                108                      -32                   34  ...   

   3 kill rounds  4 kill rounds  5 kill rounds Rifle kills Sniper kills  \
0             13              0              1          71           62   
1             12              3              0         175            0   
2              5              1              0         114            0   
3              5              2              0         145            2   
4              5              2              0         115            6   

   SMG kills  Pistol kills  Grenade  Other  \
0          0            26        1      1   
1          4            28        1      1   
2          3            16        2      0   
3          6            28        0      3   
4          7            18        6      0   

                                          Event link  
0  https://www.hltv.org/stats/players/individual/...  
1  https://www.hltv.org/stats/players/individual/...  
2  https://www.hltv.org/stats/players/individual/...  
3  https://www.hltv.org/stats/players/individual/...  
4  https://www.hltv.org/stats/players/individual/...  

[5 rows x 28 columns]

In [29]:

df3 = pd.concat([df, df2], axis=1)
df3.head()

name nationality         team  maps  rounds  KD-diff    KD  rating  \
0     ZywOo      France     Vitality    10     277       92  1.60    1.39   
1        iM     Romania  GamerLegion    12     306       68  1.34    1.35   
2     Spinx      Israel     Vitality    10     277       50  1.30    1.24   
3       NAF      Canada       Liquid     9     245       32  1.22    1.24   
4  YEKINDAR      Latvia       Liquid     9     245       30  1.19    1.20   

                       event  event_ID  ... 3 kill rounds  4 kill rounds  \
0  BLAST.tv Paris Major 2023      6973  ...          13.0            0.0   
1  BLAST.tv Paris Major 2023      6973  ...          12.0            3.0   
2  BLAST.tv Paris Major 2023      6973  ...           5.0            1.0   
3  BLAST.tv Paris Major 2023      6973  ...           5.0            2.0   
4  BLAST.tv Paris Major 2023      6973  ...           5.0            2.0   

  5 kill rounds Rifle kills  Sniper kills  SMG kills  Pistol kills  Grenade  \
0           1.0        71.0          62.0        0.0          26.0      1.0   
1           0.0       175.0           0.0        4.0          28.0      1.0   
2           0.0       114.0           0.0        3.0          16.0      2.0   
3           0.0       145.0           2.0        6.0          28.0      0.0   
4           0.0       115.0           6.0        7.0          18.0      6.0   

   Other                                         Event link  
0    1.0  https://www.hltv.org/stats/players/individual/...  
1    1.0  https://www.hltv.org/stats/players/individual/...  
2    0.0  https://www.hltv.org/stats/players/individual/...  
3    3.0  https://www.hltv.org/stats/players/individual/...  
4    0.0  https://www.hltv.org/stats/players/individual/...  

[5 rows x 41 columns]

In [30]:
df3.to_csv('salida.csv', index = False)

There are a lot of missing values, let's fix that

In [35]:
df3.tail(10)

name             nationality           team  maps  rounds  KD-diff  \
1511      adreN           United States      iBUYPOWER     2      49      -12   
1512      ultra                  Norway          Xapso     2      45      -14   
1513       pita  Bosnia and Herzegovina             SK     2      46      -16   
1514  tonyblack                  Russia  Natus Vincere     2      54      -19   
1515      ioRek                  France    Clan-Mystik     3      70      -24   
1516      HaRts                  France    Clan-Mystik     3      70      -22   
1517      xelos                  Sweden             SK     2      46      -17   
1518     starix                 Ukraine  Natus Vincere     2      54      -22   
1519     MODDII                  Sweden             SK     2      46      -22   
1520     robiin                  Sweden          Xapso     2      45      -26   

        KD  rating                  event  event_ID  ... 3 kill rounds  \
1511  0.67    0.72  DreamHack Winter 2013      1270  ...           NaN   
1512  0.58    0.65  DreamHack Winter 2013      1270  ...           NaN   
1513  0.58    0.62  DreamHack Winter 2013      1270  ...           NaN   
1514  0.59    0.60  DreamHack Winter 2013      1270  ...           NaN   
1515  0.56    0.60  DreamHack Winter 2013      1270  ...           NaN   
1516  0.55    0.59  DreamHack Winter 2013      1270  ...           NaN   
1517  0.55    0.54  DreamHack Winter 2013      1270  ...           NaN   
1518  0.50    0.52  DreamHack Winter 2013      1270  ...           NaN   
1519  0.39    0.51  DreamHack Winter 2013      1270  ...           NaN   
1520  0.26    0.40  DreamHack Winter 2013      1270  ...           NaN   

      4 kill rounds 5 kill rounds Rifle kills  Sniper kills  SMG kills  \
1511            NaN           NaN         NaN           NaN        NaN   
1512            NaN           NaN         NaN           NaN        NaN   
1513            NaN           NaN         NaN           NaN        NaN   
1514            NaN           NaN         NaN           NaN        NaN   
1515            NaN           NaN         NaN           NaN        NaN   
1516            NaN           NaN         NaN           NaN        NaN   
1517            NaN           NaN         NaN           NaN        NaN   
1518            NaN           NaN         NaN           NaN        NaN   
1519            NaN           NaN         NaN           NaN        NaN   
1520            NaN           NaN         NaN           NaN        NaN   

      Pistol kills  Grenade  Other  Event link  
1511           NaN      NaN    NaN         NaN  
1512           NaN      NaN    NaN         NaN  
1513           NaN      NaN    NaN         NaN  
1514           NaN      NaN    NaN         NaN  
1515           NaN      NaN    NaN         NaN  
1516           NaN      NaN    NaN         NaN  
1517           NaN      NaN    NaN         NaN  
1518           NaN      NaN    NaN         NaN  
1519           NaN      NaN    NaN         NaN  
1520           NaN      NaN    NaN         NaN  

[10 rows x 41 columns]

Missing values come in the form of NaN (not a number) and 0 (the player ID and event ID combination didn't exist on hltv so it displayed 0).

First let's fix the NaNs

In [41]:
df_nans = df3[df3['3 kill rounds'].isna()]
df_nans.info()

<class 'pandas.core.frame.DataFrame'>
Index: 110 entries, 1411 to 1520
Data columns (total 41 columns):
 #   Column                             Non-Null Count  Dtype  
---  ------                             --------------  -----  
 0   name                               110 non-null    object 
 1   nationality                        110 non-null    object 
 2   team                               110 non-null    object 
 3   maps                               110 non-null    int64  
 4   rounds                             110 non-null    int64  
 5   KD-diff                            110 non-null    int64  
 6   KD                                 110 non-null    float64
 7   rating                             110 non-null    float64
 8   event                              110 non-null    object 
 9   event_ID                           110 non-null    int64  
 10  Full name                          110 non-null    object 
 11  Player ID                          110 non-null    int64  


"Index: 110 entries, 1411 to 1520"

Those are the rows we need to fix

In [61]:
from selenium import webdriver
driver = webdriver.Chrome()
from bs4 import BeautifulSoup
driver.quit()

In [54]:
df_nans.columns = ['name', 'nationality', 'team', 'maps', 'rounds', 'KD-diff', 'KD',
       'rating', 'event', 'event_ID', 'Full name', 'Player ID', 'Full link',
       'name2', 'Player ID2', 'event_ID2', 'Kills', 'Deaths', 'Kill / Death',
       'Kill / Round', 'Rounds with kills', 'Kill - Death difference',
       'Total opening kills', 'Total opening deaths', 'Opening kill ratio',
       'Opening kill rating', 'Team win percent after first kill',
       'First kill in won rounds', '0 kill rounds', '1 kill rounds',
       '2 kill rounds', '3 kill rounds', '4 kill rounds', '5 kill rounds',
       'Rifle kills', 'Sniper kills', 'SMG kills', 'Pistol kills', 'Grenade',
       'Other', 'Event link']

In [67]:
Output_DF = pd.DataFrame(columns = df_nans.columns)

In [69]:
processing_list = df_nans
for index, row in processing_list.iterrows():
    Ending_index = 122
    total_rows = 122
    save_index = 0
    File_save_iter = 10

    player_id = row['Player ID']
    event_id = row['event_ID']
    player_name = row['name']

    URL = f"https://www.hltv.org/stats/players/individual/{player_id}/{player_name.lower()}?event={event_id}"
    driver = webdriver.Chrome()
    driver.get(URL)
    page_source = driver.page_source
    driver.quit()

    soup1 = BeautifulSoup(page_source, "html.parser")
    soup2 = BeautifulSoup(soup1.prettify(), "html.parser")

    stats_rows = soup2.find_all('div', class_='stats-row')
    stat_dict = {}  # Create a new empty dictionary
    print(f"[{index}/{Ending_index} | {total_rows}] Extracting data for {player_name}...")
    stat_dict['name2'] = player_name
    stat_dict['Player ID2'] = player_id
    stat_dict['event_ID2'] = event_id
    stat_dict['Event link'] = f"https://www.hltv.org/stats/players/individual/{player_id}/{player_name}?event={event_id}"
    for row in stats_rows:
        stat_name = row.find('span').text.strip()
        stat_value = row.find_all('span')[1].text.strip()
        stat_dict[stat_name] = stat_value  # Add to dictionary
    
    stat_dict['Kill - Death difference'] = int(stat_dict['Kills']) - int(stat_dict['Deaths'])

    Output_DF = pd.concat([Output_DF, pd.DataFrame([stat_dict])], ignore_index=True)
    print(f"[{index}/{Ending_index} | {total_rows}] Added statistics for Player {player_name} with ID {player_id} in event {event_id}")
    save_index += 1
    if save_index >= File_save_iter:
        Output_DF.to_csv("nans_stats.csv", index=False)
        print(f"[SAVED] > File saved after {File_save_iter} iterations, as nans_stats.csv")
        save_index = 0
    
Output_DF.to_csv("nans_stats.csv", index=False)
print("[SAVED] > File saved as nans_stats.csv")

[1411/122 | 122] Extracting data for Friis...
[1411/122 | 122] Added statistics for Player Friis with ID 7 in event 1333
[1412/122 | 122] Extracting data for GuardiaN...
[1412/122 | 122] Added statistics for Player GuardiaN with ID 2757 in event 1333
[1413/122 | 122] Extracting data for cype...
[1413/122 | 122] Added statistics for Player cype with ID 7445 in event 1333
[1414/122 | 122] Extracting data for TaZ...
[1414/122 | 122] Added statistics for Player TaZ with ID 161 in event 1333
[1415/122 | 122] Extracting data for Havoc...
[1415/122 | 122] Added statistics for Player Havoc with ID 8083 in event 1333
[1416/122 | 122] Extracting data for n0thing...
[1416/122 | 122] Added statistics for Player n0thing with ID 203 in event 1333
[1417/122 | 122] Extracting data for Hiko...
[1417/122 | 122] Added statistics for Player Hiko with ID 2644 in event 1333
[1418/122 | 122] Extracting data for SEMPHIS...
[1418/122 | 122] Added statistics for Player SEMPHIS with ID 1915 in event 1333
[1419/1